vault/src/metrics.rs

```diff
@@ -1,42 +1,249 @@
-use crate::system::VaultIdManager;
-use bitcoin::{BitcoinCoreApi, SignedAmount};
-use lazy_static::lazy_static;
+use std::{collections::HashMap, convert::TryInto, sync::Arc};
 
+use crate::system::{VaultData, VaultIdManager};
+use bitcoin::{BitcoinCoreApi, GetTransactionResultDetailCategory, SignedAmount, TransactionExt};
+use futures::{try_join, StreamExt};
+use lazy_static::lazy_static;
 use runtime::{
-    prometheus::{gather, proto::MetricFamily, Encoder, Gauge, Registry, TextEncoder},
-    CurrencyIdExt, CurrencyInfo, Error, FeedValuesEvent, FixedPointNumber,
+    prometheus::{
+        gather, proto::MetricFamily, Encoder, Gauge, GaugeVec, IntGauge, IntGaugeVec, Opts, Registry, TextEncoder,
+    },
+    CollateralBalancesPallet, CurrencyIdExt, CurrencyInfo, Error, FeedValuesEvent, FixedPointNumber,
     FixedPointTraits::One,
-    FixedU128, InterBtcParachain, OracleKey, VaultId, VaultRegistryPallet,
+    FixedU128, InterBtcParachain, IssuePallet, IssueRequestStatus, OracleKey, RedeemPallet, RedeemRequestStatus,
+    RefundPallet, ReplacePallet, UtilFuncs, VaultId, VaultRegistryPallet,
 };
 use service::{
     warp::{Rejection, Reply},
     Error as ServiceError,
 };
-use tokio::sync::RwLock;
+use std::time::Duration;
+use tokio::{sync::RwLock, time::sleep};
+const SLEEP_DURATION: Duration = Duration::from_secs(5 * 60);
+
+const CURRENCY_LABEL: &str = "currency";
+const BTC_BALANCE_TYPE_LABEL: &str = "type";
+const REQUEST_STATUS_LABEL: &str = "status";
 
-// Metrics are stored under the `vault_id` key so that multiple vaults can be easily
+// Metrics are stored under the [`CURRENCY_LABEL`] key so that multiple vaults can be easily
 // monitored at the same time.
 lazy_static! {
-    static ref BITCOIN_FEES: RwLock<Vec<f64>> = RwLock::new(vec![]);
```

Comment:
My reasoning for storing an array was that we might want to track outliers or other statistics, so I was wary of over-optimizing